### PR TITLE
Add Best practice examples

### PR DESCRIPTION
**Pull request by WestleyR**
## Description of this PR:

This PR will add a best practice/example.

I'm not the best at writing documents, but something is better then nothing :wink:

I'm *something* in here is incorrect, please comment and I will fix it!

### This PR fixed/address the following issue(s):

 - Fixes: #160

<br>

### CODE REVIEW

**Comment by WestleyR** [comment]
This PR was suggested by @stephenmathieson:
https://github.com/clibs/clib/issues/159#issuecomment-463692289


Yeah I see your point about the `src` directory, I probably should delete the directory structure code block, and make a note in the Makefile for somthing...

*BTW, I'm very new to clib, less then a week, so my knowledge about clib is kinda low... for now.*

File: BEST_PRACTICE.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+ - [How to install/uninstall executables](#install-and-uninstall-executables).
+
+For instructions on installation, click [here](https://github.com/clibs/clib#installation).
+
+## How to use installed libraries for your project
+
+Lets say you have your project, with a typical directory tree like this:
+
+```
+your-project/
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Please use the `tree` program to generate this. It looks much "nicer" 😄

File: BEST_PRACTICE.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+├── src/
+│   ├── main.c
+│   ├── function.c
+│   ├── function.h
+│   └── etc...
+│
+└── test.sh
+```
+
+`test.sh` is your test script. `src` is the directory where you source
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Not many projects use a `test.sh` script. Tests are more commonly written in C.

File: BEST_PRACTICE.md [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+CC = gcc
+
+# where to install
+PREFIX = /usr/local/bin
+
+# your output file
+TARGET = project-name
+
+# your project name
+TARGET = your-project
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Duplicate?

File: BEST_PRACTICE.md [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+.PHONY:
+install: $(TARGET)
+	cp -f $(TARGET) $(PREFIX)
+
+.PHONY:
+uninstall: $(PREFIX)/$(TARGET)
+	rm -f $(PREFIX)/$(TARGET)
+```
+
+This is a basic Makefile, and should work for most of your projects.
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Might be nice to just link to an existing Makefile for this, or use a Gist.

File: BEST_PRACTICE.md [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+
+You *could* have your Makefile install the libraries upon running it, but you
+would only need to do that to get the latest version of the library(s), in this
+case you proboly don't want that. You typicly want yo get the lates stable version
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Several typos in this paragraph:

- `s/proboly/probably/`
- `s/typicly/typically/`
- `s/lates/latest/`

Also, please remove the stray newlines.

File: BEST_PRACTICE.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+for that library. By having a `package.json` file in your project repo, you can
+specify what packages you need, and what version of that package. Now have a look
+at a example `package.json` file for your project: (executable package)
+
+### Example package.json for exacutable project
+
+```
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Add `json` here.

File: BEST_PRACTICE.md [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+    "stephenmathieson/trim.c": "0.0.2",
+    "clibs/commander": "1.3.2",
+    "clibs/logger": "0.0.1",
+  },
+  "install": "make install",
+  "uninstall": "make uninstall"
+}
+```
+
+Starting from the top, `"name"` is your package name. `"version"` is your package version.
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Please remove the stray newlines.

File: BEST_PRACTICE.md [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+`"repo"` is the location of your project, (not including the `https://github.com/`).
+`"dependencies"` is all the dependencies your project requiers, along with there version.
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: "requiers" should be "requires".

File: BEST_PRACTICE.md [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+`"install"` is the command to install your program (ran as root), (tip: if your project
+requiers more then one command to install it, like need to run `./configure`, before `make`,
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Typo.

File: BEST_PRACTICE.md [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+then do this: `"install": "./configure && make && make install"`). `"uninstall"` is the command
+to uninstall your project, [more on that later](#install-and-uninstall-executables).
+
+***NOTE:** Make sure you have a relese as the same version in your `package.json` file, otherwise
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Please remove stray newlines.

Comment (stephenmathieson) [inline BEST_PRACTICE.md]: This doesn't look like valid Markdown. Why are there so many `*`s here?

File: BEST_PRACTICE.md [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+│
+├── src/
+│   ├── library.c
+│   ├── library.h
+│   └── etc...
+│
+└── test.sh
+```
+
+Also like before, your have a `deps` directory (depending on your library, you may not need any
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Why is this bolded?

File: BEST_PRACTICE.md [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+  }
+  "license": "YOUR_LIB LICENSE",
+  "src": [
+    "src/library.c",
+    "src/library.h"
+  ],
+}
+```
+
+The main difrances (between this, and the executable `package.json`), is now there is `"src"`,
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Differences typo.

File: BEST_PRACTICE.md [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+this is where your make library source code is, your can change it, but src is pritty standard.
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: Pretty typo.

File: BEST_PRACTICE.md [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+After you install `clib-uninstall` you can use it like so:
+
+```bash
+$ sudo clib-uninstall <username/package-name>
+
+# for example:
+$ sudo clib-uninstall visionmedia/mon
+```
+
+<br>
```
Comment (stephenmathieson) [inline BEST_PRACTICE.md]: What is this for?

File: Readme.md [hunk 15]
```diff
@@ -68,6 +68,8 @@ $ sudo make install
 
 ## Examples
 
+ For full example/best practice, click [here](https://github.com/clibs/clib/blob/master/BEST_PRACTICE.md).
```
Comment (stephenmathieson) [inline Readme.md]: Maybe reword this as:

> More examples and best practices at BEST_PRACTICE.md

IMO it reads a bit better and [doesn't have a "click here" link](https://webaccess.berkeley.edu/ask-pecan/click-here).

File: BEST_PRACTICE.md (hunk 1 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: I did use the `tree` program, I just copy-pasted it. I dont know why it looks different here :thinking:

File: BEST_PRACTICE.md (hunk 2 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: Okay, I'll remove it :+1:

File: BEST_PRACTICE.md (hunk 3 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: Opps!

File: BEST_PRACTICE.md (hunk 4 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: I'll be fine with that, although, if I was a user, I prefer to see the Makefile right here. But if you have a Gist Makefile example then let me know. _it should be under your user, rather then mine :wink:_

File: BEST_PRACTICE.md (hunk 5 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: Dam! thats my record for typos!

> Also, please remove the stray newlines.

Where? Do you mean the newlines that breakup this paragraph into line no longer then 80 chars? I did that so its easy to read, and suggest changes. It wont make a difference in the final markdown file.

File: BEST_PRACTICE.md (hunk 6 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: Duh! :+1:

File: BEST_PRACTICE.md (hunk 7 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: :frowning_face: :frowning_face: :frowning_face: :frowning_face: Fine, but its easier to read and suggest changes like this, the final markdown file will look the same when viewing it.

File: BEST_PRACTICE.md (hunk 8 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: :+1:

File: BEST_PRACTICE.md (hunk 9 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: :+1:

File: BEST_PRACTICE.md (hunk 10 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: :+1:

Comment (WestleyR) [reply BEST_PRACTICE.md]: I'll try to fix that...

File: BEST_PRACTICE.md (hunk 11 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: Its not, Github is not printing it correctly with all the `***` from above. I'll try to fix that...

File: BEST_PRACTICE.md (hunk 12 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: :+1:

File: BEST_PRACTICE.md (hunk 13 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: :+1:

File: BEST_PRACTICE.md (hunk 14 above)
Comment (WestleyR) [reply BEST_PRACTICE.md]: Because I'm weird. Just a break at the end of the file, not necessary.

**Comment by Isty001** [comment]
What's the state of this PR? 

@WestleyR could you look at the review sometime please?

**Comment by WestleyR** [comment]
> What's the state of this PR?
> 
> @WestleyR could you look at the review sometime please?

Yep, should be good for another review.
